src/backend/distributed/connection/remote_commands.c

```diff
@@ -652,3 +656,224 @@ FinishConnectionIO(MultiConnection *connection, bool raiseInterrupts)
+
+			/*
+			 * We cannot disable wait events as of postgres 9.6, so we rebuild the
+			 * WaitEventSet whenever connections are ready.
+			 */
+			if (rebuildWaitEventSet)
+			{
+				if (waitEventSet != NULL)
+				{
+					FreeWaitEventSet(waitEventSet);
+				}
+			
+				waitEventSet = BuildWaitEventSet(allConnections, totalConnectionCount,
+												 pendingConnectionsStartIndex);
+
+				rebuildWaitEventSet = false;
+			}
+
+			/* wait for I/O events */
+#if (PG_VERSION_NUM >= 100000)
+			eventCount = WaitEventSetWait(waitEventSet, timeout, events,
+										  pendingConnectionCount, WAIT_EVENT_CLIENT_READ);
+#else
+			eventCount = WaitEventSetWait(waitEventSet, timeout, events,
+										  pendingConnectionCount);
+#endif
+
+			/* process I/O events */
+			for (; eventIndex < eventCount; eventIndex++)
+			{
+				WaitEvent *event = &events[eventIndex];
+				MultiConnection *connection = NULL;
+				bool connectionIsReady = false;
+
+				if (event->events & WL_POSTMASTER_DEATH)
+				{
+					ereport(ERROR, (errmsg("postmaster was shut down, exiting")));
+				}
+
+				if (event->events & WL_LATCH_SET)
+				{
+					ResetLatch(MyLatch);
+
+					if (raiseInterrupts)
+					{
+						CHECK_FOR_INTERRUPTS();
+					}
+
+					if (InterruptHoldoffCount > 0 && (QueryCancelPending || ProcDiePending))
+					{
+						/* return immediately in case of cancellation */
+						FreeWaitEventSet(waitEventSet);
+						return;
+					}
+
+					continue;
+				}
+
+				connection = (MultiConnection *) event->user_data;
+				connectionIndex = event->pos + pendingConnectionsStartIndex;
+
+				if (event->events & WL_SOCKET_WRITEABLE)
+				{
+					int sendStatus = PQflush(connection->pgConn);
+					if (sendStatus == -1)
+					{
+						/* send failed, done with this connection */
+						connectionIsReady = true;
+					}
+					else if (sendStatus == 0)
+					{
+						/* done writing, only wait for read events */
+						ModifyWaitEvent(waitEventSet, connectionIndex, WL_SOCKET_READABLE,
+										NULL);
+					}
+				}
+
+				if (event->events & WL_SOCKET_READABLE)
+				{
+					int receiveStatus = PQconsumeInput(connection->pgConn);
+					if (receiveStatus == 0)
+					{
+						/* receive failed, done with this connection */
+						connectionIsReady = true;
+					}
+					else if (!PQisBusy(connection->pgConn))
+					{
+						/* result was received */
+						connectionIsReady = true;
+					}
+				}
+
+				if (connectionIsReady)
+				{
+					connectionReady[connectionIndex] = true;
+					rebuildWaitEventSet = true;
+				}
+			}
+
+			/* move non-ready connections to the back of the array */
+			for (connectionIndex = pendingConnectionsStartIndex;
+				 connectionIndex < totalConnectionCount; connectionIndex++)
+			{
+				if (connectionReady[connectionIndex])
+				{
+					allConnections[connectionIndex] = allConnections[pendingConnectionsStartIndex];
+					pendingConnectionsStartIndex++;
+				}
+			}
+
+		}
+
+		if (waitEventSet != NULL)
+		{
+			FreeWaitEventSet(waitEventSet);
```

Comment:
probably good policy to reset waitEventSet to NULL afterwards, both here and below. Quite possible that somebody adds code that can error out afterwards...